### PR TITLE
Port form data parsing to multipart

### LINKED ISSUE #39: cgi is going to be deprecated in Python 3.8 and removed in 3.10
zope.publisher relies on stdlib's `cgi.FieldStorage()` for HTTP request parsing.  [PEP-594 indicates that `cgi` will be removed in Python 3.10](https://www.python.org/dev/peps/pep-0594/#cgi).  What do we want to do?

- Rewrite HTTP request parsing
- Ask Python core developers nicely to please keep `cgi` (or maybe just `cgi.FieldStorage`) alive longer
- Something else?

### PR DESCRIPTION
**Pull request by cjwatson**
This currently adds ResourceWarnings on Python 3; that will be fixed
when https://github.com/defnull/multipart/pull/21 lands.

Fixes #39.
Closes #54.

See also my alternative solution in #54.  (I prefer this one.)

### CODE REVIEW

File: src/zope/publisher/browser.py
Comment (mgedmin) [inline src/zope/publisher/browser.py:301]: I'm not sure I understand this comment.  Does this mean "cgi.FieldStorage used to set Content-Type if it was missing, so we do that here, now that we don't use cgi.FieldStorage any more"?

File: src/zope/publisher/browser.py
Comment (mgedmin) [inline src/zope/publisher/browser.py:336]: This seems frighteningly non-thread-safe.

Thread 1: _orig_temporary_file = multipart's TemporaryFile
Thread 1: multipart.TemporaryFile = tempfile.NamedTemporaryFile
Thread 2: _orig_temporary_file = multipart.TemporaryFile = tempfile.NamedTemporaryFile
Thread 1. multipart.TemporaryFile = _orig_temporary_file = multipart's TemporaryFile
Thread 2: multipart.TemporaryFile = _orig_temporary_file = tempfile.NamedTemporaryFile

and now multipart.TemporaryFile is always and forever tempfile.NamedTemporaryFile.

Seems less bad to just monkey-patch the thing at import time, if we must monkeypatch.

(also, I'm curious: under what circumstances can multipart.TemporaryFile be None?)

File: src/zope/publisher/browser.py
Comment (mgedmin) [inline src/zope/publisher/browser.py:641]: Maybe add a compatibility alias `ZopeFieldStorage = FieldStorage`, in case something imports and uses it?

(Nothing in zopefoundation/ does this, but there might packages elsewhere.  z3c.soap does import ZopeFieldStorage, but from ZPublisher.HTTPRequest via a star import.)

File: src/zope/publisher/browser.py
Comment (mgedmin) [inline src/zope/publisher/browser.py:322]: Uh, oh, the M word.

**Review by mgedmin — CHANGES_REQUESTED** [review]
The monkeypaching gives me pause.  The rest looks good.

File: src/zope/publisher/browser.py
Comment (cjwatson) [reply src/zope/publisher/browser.py]: Yes, exactly.  I've clarified that comment.

File: src/zope/publisher/browser.py
Comment (cjwatson) [reply src/zope/publisher/browser.py:337]: The guard against it being None was to defend against future changes to multipart that switched to `NamedTemporaryFile`, which at the time I thought might be a plausible change (though multipart's maintainer has a reasonable point about that not working on Windows).

Good point on thread-safety.  What do you think about just taking the API break and dropping all this monkey-patching?  I'm not really sure what the point of exposing the `name` attribute is; it can't be used in portable code, it doesn't convey any information about the form since it's just a randomly-generated path, and `FileUpload` exposes all the file object methods that a caller would need to read the file's contents.  And in https://github.com/defnull/multipart/pull/22 the original author of this didn't really seem to know why it was there.

File: src/zope/publisher/browser.py
Comment (cjwatson) [reply src/zope/publisher/browser.py:641]: I've added a `from cgi import FieldStorage as ZopeFieldStorage`, ignoring `ImportError`s.  Is that the sort of thing you meant?

File: src/zope/publisher/browser.py
Comment (mgedmin) [reply src/zope/publisher/browser.py:341]: For `None` to work as a guard I think you need to up update the `getattr` call and provide None as a default value.

(Oops: I looked at the getattr with a constant attribute literal and no default value and didn't wonder why you used getattr instead of regular attribute syntax.)

Comment (mgedmin) [reply src/zope/publisher/browser.py:341]: I don't know of any code that expects Zope form file fields to have a `name` attribute.  (This doesn't mean such code doesn't exist, just I'm not aware of any.)

An API break seems provisionally acceptable to me.  We can always wait and see if anybody complains, and then we'll have a use case we can take upstream to `multipart`.

File: src/zope/publisher/browser.py
Comment (mgedmin) [reply src/zope/publisher/browser.py:641]: Ok, I think what happened was I was under the misapprehension that the code here subclassed cgi.FieldStorage into a FieldStorage and then had a 2nd subclass as ZopeFieldStorage, for the purposes of I don't even know what.

I'm -1 for making ZopeFieldStorage available as an alias for cgi.FieldStorage if this changes behavior of ZopeFieldStorage from previous zope.publisher versions.  Let's just drop it.

File: src/zope/publisher/browser.py
Comment (cjwatson) [reply src/zope/publisher/browser.py]: OK, bumped to 6.0.0.

**Comment by cjwatson** [comment]
I'm going to leave this a week or so to see if anyone has other thoughts, and then land this in the absence of objections.

**Comment by cjwatson** [comment]
I ran this though Launchpad's test suite, which caught a couple of mistakes.  I've pushed some fixes for those and would appreciate re-review.

It may be necessary to wait for defnull/multipart#25 and a Python 2 equivalent before landing this.  Without this, Launchpad's test suite regresses in a way that would be hard to work around in zope.publisher.  I don't know how widespread the effect of this would be on other users of zope.publisher; it affects us due to what's arguably a bug in wadllib, but that bug was inherited from using the standard library's `email.mime` to construct `multipart/form-data` request bodies, so it's possible others have made a similar mistake.

**Comment by icemac** [comment]
Everybody voted for this PR over #54. @mgedmin Could you please re-review the last changes so it can eventually be merged?

**Comment by icemac** [comment]
defnull/multipart#21 has landed in the release 0.2.2 of multipart.

**Review by mgedmin — APPROVED** [review]
lgtm!

**Comment by cjwatson** [comment]
I agree this is preferable to #54, but just note that we still need defnull/multipart#25 to land (and a Python 2 equivalent) first.  The multipart maintainer offered me co-maintenance which I'm happy to do, but I'm waiting for the actual corresponding permission bits before I can move forward with that.

**Comment by jugmac00** [comment]
> I agree this is preferable to #54, but just note that we still need [defnull/multipart#25](https://github.com/defnull/multipart/pull/25) to land (and a Python 2 equivalent) first. The multipart maintainer offered me co-maintenance which I'm happy to do, but I'm waiting for the actual corresponding permission bits before I can move forward with that.

That is great news! I was a bit worried to rely on an external package, which only one person has access to, and is not seemingly not very active. But with you being a co-maintainer, this sounds excellent! Thanks for all your work you already put into this issue!

**Comment by icemac** [comment]
https://github.com/defnull/multipart/pull/25 Seems to be merged. What are the next steps?

**Comment by cjwatson** [comment]
I've released multipart 0.1.1 and 0.2.3 now, so I'm going to go ahead and land this.  Thanks all.

**Pull request merged**
